File: i_pylox/ast_printer.py
import expr
from token import Token, TokenType

class AstPrinter(expr.ExprVisitor):

    def print(self, expr):
        return expr.accept(self)

    def visitBinaryExpr(self, expr):
        return self.parenthesize(expr.operator.lexeme, expr.left, expr.right)

    def visitGroupingExpr(self, expr):
        return self.parenthesize("group", expr.expression)

    def visitLiteralExpr(self, expr):
        if expr.value is None: return "nil"
        return str(expr.value)

    def visitUnaryExpr(self, expr):
        return self.parenthesize(expr.operator.lexeme, expr.right)

    def parenthesize(self, name, *exprs):
        s = '(' + name
        for expr in exprs:
            s += ' '
            s += expr.accept(self)
        s += ')'

        return s

if __name__ == '__main__':
    e = expr.Binary(
        expr.Unary(
            Token(TokenType.MINUS, '-', None, 1),
            expr.Literal(123)),
        Token(TokenType.STAR, '*', None, 1),
        expr.Grouping(expr.Literal(45.67)))

    print(AstPrinter().print(e))

File: jlox/src/main/java/ahodanenok/craftinginterpeters/lox/AstPrinter.java
package ahodanenok.craftinginterpreters.lox;

public class AstPrinter implements Expression.Visitor<String> {

    public static void main(String... args) {
        Expression expression =
            new Expression.Binary(
                new Token(TokenType.STAR, "*", null, 1),
                new Expression.Unary(
                    new Token(TokenType.MINUS, "-", null, 1),
                    new Expression.Literal(123)),
                new Expression.Grouping(
                    new Expression.Literal(45.67)));
        System.out.println(new AstPrinter().print(expression));
    }

    String print(Expression expression) {
        return expression.accept(this);
    }

    @Override
    public String visitLiteralExpression(Expression.Literal expression) {
        if (expression.value != null) {
            return expression.value.toString();
        } else {
            return "nil";
        }
    }

    @Override
    public String visitUnaryExpression(Expression.Unary expression) {
        return parethesize(expression.operator.lexeme, expression.expression);
    }

    @Override
    public String visitBinaryExpression(Expression.Binary expression) {
        return parethesize(expression.operator.lexeme, expression.left, expression.right);
    }

    @Override
    public String visitTernaryExpression(Expression.Ternary expression) {
        return parethesize("?:", expression.condition, expression.left, expression.right);
    }

    @Override
    public String visitGroupingExpression(Expression.Grouping expression) {
        return parethesize("group", expression.expression);
    }

    @Override
    public String visitVariableExpression(Expression.Variable expression) {
        return String.format("(variable %s)", expression.name.lexeme);
    }

    @Override
    public String visitAssignExpression(Expression.Assign expression) {
        return String.format("(assign %s %s)", expression.name.lexeme, expression.expression.accept(this));
    }

    @Override
    public String visitLogicalExpression(Expression.Logical expression) {
        return parethesize(expression.operator.lexeme, expression.left, expression.right);
    }

    private String parethesize(String type, Expression... expressions) {
        StringBuilder sb = new StringBuilder();
        sb.append("(").append(type);
        for (Expression expression : expressions) {
            sb.append(" ").append(expression.accept(this));
        }
        sb.append(")");

        return sb.toString();
    }
}
